test/test_jit.py

```diff
@@ -15035,6 +15035,53 @@ def _xor():  # noqa: E306
             def test():
                 return Foo(torch.tensor(1)) + Foo(torch.tensor(1))
 
+    def test_cast_overloads(self):
+        @torch.jit.script
+        class Foo(object):
+            def __init__(self, val):
+                # type: (float) -> None
+                self.val = val
+
+            def __int__(self):
+                return int(self.val)
+
+            def __float__(self):
+                return self.val
+
+            def __bool__(self):
+                return bool(self.val)
+
+            def __str__(self):
+                return str(self.val)
+
+        def test(foo):
+            # type: (Foo) -> Tuple[int, float, bool]
+            if foo:
+                pass
+            return int(foo), float(foo), bool(foo)
+
+        fn = torch.jit.script(test)
+        self.assertEqual(fn(Foo(0.5)), test(0.5))
+        self.assertEqual(fn(Foo(0.)), test(0.0))
+        # str has slightly different formatting
+        self.assertTrue("0.5" in (str(Foo(0.5))))
+        self.assertTrue("0." in (str(Foo(0.0))))
+
+        @torch.jit.script
+        class BadBool(object):
+            def __init__(self):
+                pass
+
+            def __bool__(self):
+                return (1, 2)
+
+        with self.assertRaisesRegex(RuntimeError, "expected a bool expression for condition"):
```

Comment:
The error here is up a level too high, it should be thrown at the cast level to match Python:

```python
class X(object):
    def __bool__(self):
        return (1, 2)

x = X()
if x:
    print("hi")
```
prints
```
Traceback (most recent call last):
  File "../test.py", line 142, in <module>
    if x:
TypeError: __bool__ should return bool, returned tuple
```

Reply:
We don't currently have the infrastructure for that. This could be possible with Interfaces, but is not currently possible. I don't think that should be a blocker for landing this.

Reply:
In CastValue you know the type it should be and what the type actually is, can't you compare there?

Reply:
That doesn't fit in with MagicMethod sugared val, which abstracts over the input sugared val. Where would that check go?

Reply:
`MagicMethod` calls `CastValue`, so it'd be fine there, then you could also get rid of the special erroring for `emitCond`

Reply:
MagicMethod also calls BuiltinFunctions - it doesn't just call CastValue. it's an abstraction